packages/labs/router/src/routes.ts

```diff
@@ -0,0 +1,307 @@
+   * changes may result is a different route matching for the current path, you
+   * must instigate a route update with `goto()`.
+   */
+  routes: Array<RouteConfig> = [];
+
+  /*
+   * The current set of child Routes controllers. These are connected via
+   * the routes-connected event.
+   */
+  private readonly _childRoutes: Array<Routes> = [];
+
+  protected _parentRoutes: Routes | undefined;
+
+  /*
+   * State related to the current matching route.
+   *
+   * We keep this so that consuming code can access current parameters, and so
+   * that we can propagate tail matches to child routes if they are added after
+   * navigation / matching.
+   */
+  protected _currentPathname: string | undefined;
+  protected _currentRoute: RouteConfig | undefined;
+  protected _currentParams: {
+    [key: string]: string;
+  } = {};
+
+  /**
+   * Callback to call when this controller is disconnected.
+   *
+   * It's critical to call this immediately in hostDisconnected so that this
+   * controller instance doesn't receive a tail match meant for another route.
+   */
+  // TODO (justinfagnani): Do we need this now that we have a direct reference
+  // to the parent? We can call `this._parentRoutes.disconnect(this)`.
+  private _onDisconnect: (() => void) | undefined;
+
+  constructor(
+    host: ReactiveControllerHost & HTMLElement,
+    routes: Array<RouteConfig>
+  ) {
+    (this._host = host).addController(this);
+    this.routes = [...routes];
+  }
+
+  /**
+   * Returns a URL string of the current route, including parent routes,
+   * optionally replacing the local path with `pathname`.
+   */
+  link(pathname?: string): string {
+    if (pathname?.startsWith('/')) {
+      return pathname;
+    }
+    if (pathname?.startsWith('.')) {
+      throw new Error('Not implemented');
+    }
+    pathname ??= this._currentPathname;
+    return (this._parentRoutes?.link() ?? '') + pathname;
+  }
+
+  /**
+   * Navigates this routes controller to `pathname`.
+   *
+   * This does not navigate parent routes, so it isn't (yet) a general page
+   * navigation API. It does navigate child routes if pathname matches a
+   * pattern with a tail wildcard pattern (`/*`).
+   */
+  async goto(pathname: string) {
+    // TODO (justinfagnani): handle absolute vs relative paths separately.
+    // TODO (justinfagnani): do we need to detect when goto() is called while
+    // a previous goto() call is still pending?
+
+    // TODO (justinfagnani): generalize this to handle query params and
+    // fragments. It currently only handles path names because it's easier to
+    // completely disregard the origin for now. The click handler only does
+    // an in-page navigation if the origin matches anyway.
+    let tailGroup: string | undefined;
+    if (this.routes.length === 0) {
+      // If a routes controller has none of its own routes it acts like it has
+      // one route of `/*` so that it passes the whole pathname as a tail
+      // match.
```

Comment:
I think this works actually, because a fallback that loads from the server should return false in it's enter, which will prevent any update to router state - those updates will instead come from the server-fetched route. I have a test that simulates that for one level, we can expand for multiple levels.